internal/k6runner/k6runner.go

```diff
@@ -328,29 +344,98 @@ type RunResponse struct {
 }
 
 func (r HttpRunner) WithLogger(logger *zerolog.Logger) Runner {
-	return HttpRunner{
-		url:    r.url,
-		logger: logger,
-	}
+	r.logger = logger
+
+	return r
 }
 
 var ErrUnexpectedStatus = errors.New("unexpected status code")
 
 func (r HttpRunner) Run(ctx context.Context, script Script) (*RunResponse, error) {
-	k6Timeout := time.Duration(script.Settings.Timeout) * time.Millisecond
+	if r.backoff == 0 {
+		panic("zero backoff, runner is misconfigured, refusing to DoS")
+	}
+
+	if deadline, hasDeadline := ctx.Deadline(); !hasDeadline {
+		defaultAllRetriesTimeout := time.Duration(script.Settings.Timeout) * time.Millisecond * 2
+		r.logger.Error().
+			Dur("allRetriesTimeout", defaultAllRetriesTimeout).
+			Msg("k6 runner does not have a deadline for all retries. This is a bug. Defaulting to twice the timeout to avoid retrying forever")
+
+		var cancel context.CancelFunc
+		ctx, cancel = context.WithTimeout(ctx, defaultAllRetriesTimeout)
+		defer cancel()
+	} else if tud := time.Until(deadline); tud < time.Duration(script.Settings.Timeout)*time.Millisecond*2 {
+		r.logger.Debug().
+			Str("timeUntilNext", tud.String()).
+			Str("timeout", (time.Duration(script.Settings.Timeout) * time.Millisecond).String()).
+			Msg("time until next execution is too close to script timeout, there might not be room for retries")
+	}
+
+	wait := r.backoff
+	var response *RunResponse
+	for {
+		start := time.Now()
+
+		var err error
+		response, err = r.request(ctx, script)
+		if err == nil {
+			r.logger.Debug().Bytes("metrics", response.Metrics).Bytes("logs", response.Logs).Msg("script result")
+			return response, nil
+		}
+
+		if !errors.Is(err, errRetryable) {
+			return nil, err
+		}
+
+		// Wait, but subtract the amount of time we've already waited as part of the request timeout.
+		// We do this because these requests have huge timeouts, and by the nature of the system running these request,
+		// we expect the most common error to be a timeout, so we avoid waiting even more on top of an already large
+		// value.
+		waitRemaining := max(0, wait-time.Since(start))
+		r.logger.Debug().Err(err).Dur("after", waitRemaining).Msg("retrying retryable error")
+
+		waitTimer := time.NewTimer(waitRemaining)
+		select {
+		case <-ctx.Done():
+			waitTimer.Stop()
+			return nil, fmt.Errorf("cannot retry further: %w", errors.Join(err, ctx.Err()))
+		case <-waitTimer.C:
+		}
+
+		// Backoff linearly, adding some jitter.
+		wait += r.backoff + time.Duration(rand.Intn(int(r.backoff)))
+	}
+}
+
+// errRetryable indicates that an error is retryable. It is typically joined with another error.
+var errRetryable = errors.New("retryable")
+
+func (r HttpRunner) request(ctx context.Context, script Script) (*RunResponse, error) {
+	checkTimeout := time.Duration(script.Settings.Timeout) * time.Millisecond
+	if checkTimeout == 0 {
+		return nil, ErrNoTimeout
+	}
```

Comment:
The only reason is that this is the place where this value is used (to calculate more stuff later). We could preemptively check it, but is there a reason to? We'd make the code more verbose.

Reply:
I see, that makes sense, I was just considering the "extra work" done by the upper level `Run` method, such as starting the loop, declaring the timer etc. But since this should happen rather infrequently and the cost is not high I guess it is fine.